File: python/hackerrank/simulation/biggerIsGreater.py

def biggerIsGreater(word):
    chars = list(word)
    codes = [ord(c) for c in chars]

    for i in range(len(codes) - 1, 0, -1):
        if codes[i] > codes[i - 1]:
            pivot = i - 1
            pivot_val = codes[pivot]
            break
    else:
        return "no answer"

    swap_idx = None
    swap_val = None

    for j in range(pivot + 1, len(codes)):
        if codes[j] > pivot_val and (swap_val is None or codes[j] < swap_val):
            swap_val = codes[j]
            swap_idx = j

    chars[pivot], chars[swap_idx] = chars[swap_idx], chars[pivot]
    chars[pivot + 1:] = sorted(chars[pivot + 1:])

    return "".join(chars)

def main():
    test_cases = int(input().strip())
    for _ in range(test_cases):
        word = input()
        print(biggerIsGreater(word))

if __name__ == "__main__":
    main()

File: golang/hackerrank/simulation/biggerIsGreater.go
package simulation

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

// BiggerIsGreater finds the smallest lexicographically greater string.
func BiggerIsGreater(word string) string {
	chars := []rune(word)
	n := len(chars)
	pivot := -1

	for i := n - 1; i > 0; i-- {
		if chars[i] > chars[i-1] {
			pivot = i - 1
			break
		}
	}

	if pivot == -1 {
		return "no answer"
	}

	swapIdx := -1
	for j := n - 1; j > pivot; j-- {
		if chars[j] > chars[pivot] {
			swapIdx = j
			break
		}
	}

	chars[pivot], chars[swapIdx] = chars[swapIdx], chars[pivot]

	// Sort suffix
	suffix := chars[pivot+1:]
	sort.Slice(suffix, func(i, j int) bool {
		return suffix[i] < suffix[j]
	})

	return string(chars)
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	lineT, _ := reader.ReadString('\n')
	tCount, _ := strconv.Atoi(strings.TrimSpace(lineT))

	for i := 0; i < tCount; i++ {
		word, _ := reader.ReadString('\n')
		fmt.Println(BiggerIsGreater(strings.TrimSpace(word)))
	}
}
